# Flight Price Prediction 
**OBJECTIVE:**
* Predict flight ticket prices based on date, destination, and other factors. 



**DATA COLLECTION:**
* Data is collected from Kaggle: Flight Price Prediction uploaded by Shubham Bathwal.

In [19]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [20]:
df=pd.read_csv('Clean_Dataset.csv')
df.info()
df.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300153 entries, 0 to 300152
Data columns (total 12 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Unnamed: 0        300153 non-null  int64  
 1   airline           300153 non-null  object 
 2   flight            300153 non-null  object 
 3   source_city       300153 non-null  object 
 4   departure_time    300153 non-null  object 
 5   stops             300153 non-null  object 
 6   arrival_time      300153 non-null  object 
 7   destination_city  300153 non-null  object 
 8   class             300153 non-null  object 
 9   duration          300153 non-null  float64
 10  days_left         300153 non-null  int64  
 11  price             300153 non-null  int64  
dtypes: float64(1), int64(3), object(8)
memory usage: 27.5+ MB


Unnamed: 0   airline   flight source_city departure_time stops  \
0           0  SpiceJet  SG-8709       Delhi        Evening  zero   
1           1  SpiceJet  SG-8157       Delhi  Early_Morning  zero   
2           2   AirAsia   I5-764       Delhi  Early_Morning  zero   
3           3   Vistara   UK-995       Delhi        Morning  zero   
4           4   Vistara   UK-963       Delhi        Morning  zero   

    arrival_time destination_city    class  duration  days_left  price  
0          Night           Mumbai  Economy      2.17          1   5953  
1        Morning           Mumbai  Economy      2.33          1   5953  
2  Early_Morning           Mumbai  Economy      2.17          1   5956  
3      Afternoon           Mumbai  Economy      2.25          1   5955  
4        Morning           Mumbai  Economy      2.33          1   5955

no null value

In [21]:
# Let's check for duplicates
df.duplicated().sum()
# no duplicates

np.int64(0)

# DISCOVER
* our data has columns like airline, flight, source_city, stops, arrivale time etc among them source and destination city data doesnot conclude anything. 
* also duration and days left tells; duration: total time of flight measure in hours and days lefts means days remaining until the flight's departure. 
* value 1 for days remaining mean flights are booked just one day before their departure.
* duration like 2.17 mean that the flights are direct flights whereas duration like 9.67 or 26.42 hr means there were layovers(waiting of flights at places or airports) before reaching mumbai.
* identify the data ranges, mean, medians and standard deviations for numerical columns

# STRUCTURING 
* organize: reshaping data, handling missing values, converting data types
* ensuring data is in standardized manner which helps for further manipulation and exploration

# CLEANING
* to address inconsistencies, errors and outliers within dataset
* correcting inaccurate data, handling duplicates, and addressing any anomalies
* focus on enhancing data accuracy and make the data error-free as possible

# ENRICHING
* adding additional datas to provide more depth or context
- mergings datasets or incorporating external data sources

# VALIDATING
- checks the quality and reliability of processed data
- by using visualization tools(boxplots, histogram, scatterplots)


# PUBLISHING

**INSHORT**
- missing values(done)
- duplicate values(done)
- correct data types
- outliers(next task to do)
- standardize/ normalize values
- typos or formatting issues
- remove irrelevant columns(done)
- encode categorical variables( label encoding/one hot encoding)
- remove invalid or unusual values(eg: datas entered range is 5-10 but the entry is 100 then remove or correct it)
- feature engineering(add/ modify the colums or features)
- check the datas and then save the cleaned datas



hamile dataset ko columns lai pani define garnu parxa

jati ota warning aauxa program ma teslai ignore garne
# Supress warnings
import warnings
warnings.filterwarnings('ignore')

print(df.shape)
(89,10) ie 89 rows and 10 columns

NOTES KO EXAMPLES:
- hamile titanic dataset leko xam, tyo vanya titanic ma bhako passenger ko age, class,etc.



In [22]:
df.duplicated().sum()

np.int64(0)

no duplicate values

- For proper data analysis, converting non-numeric features to numeric values is often necessary, but not always mandatory.
- two types: categorical and numeric.
- Categorical values need to be converted into a numerical format before they can be used in machine learning models. This process is known as encoding.
- Label encoding assigns a unique integer to each category.For categorical variable is ordinal (i.e., has a meaningful order).
- One-hot encoding creates new columns indicating the presence (or absence) of each possible value in the original data.For nominal categorical variables (i.e., no intrinsic ordering).

yo encoding ko chai benefits k hunxa?

In [26]:
df.columns

Index(['Unnamed: 0', 'airline', 'flight', 'source_city', 'departure_time',
       'stops', 'arrival_time', 'destination_city', 'class', 'duration',
       'days_left', 'price'],
      dtype='object')

In [27]:
df.rename(columns={'Unnamed: 0':'Index'},inplace=True)

In [28]:
df.columns

Index(['Index', 'airline', 'flight', 'source_city', 'departure_time', 'stops',
       'arrival_time', 'destination_city', 'class', 'duration', 'days_left',
       'price'],
      dtype='object')

object datatype means any kind of datatypes

In [29]:
df.dtypes[df.dtypes == "object"].index

Index(['airline', 'flight', 'source_city', 'departure_time', 'stops',
       'arrival_time', 'destination_city', 'class'],
      dtype='object')